File: Developer/suoke_life/services/user-service/src/repositories/user.repository.js
/**
 * 用户存储库
 */
const { v4: uuidv4 } = require('uuid');
const { db } = require('../utils/db');
const { redis, getCache, deleteCache, deleteCacheByPattern } = require('../utils/redis');
const { logger } = require('@suoke/shared').utils;
const { userModel } = require('../models');

class UserRepository {
  constructor() {
    this.tableName = userModel.TABLE_NAME;
    this.cacheKeyPrefix = 'user:';
  }

  /**
   * 根据ID获取用户
   */
  async getById(id) {
    try {
      const cacheKey = `${this.cacheKeyPrefix}id:${id}`;
      return getCache(cacheKey, async () => {
        const user = await db(this.tableName).where({ id }).first();
        return user || null;
      });
    } catch (error) {
      logger.error('获取用户失败', { error: error.message, id });
      throw error;
    }
  }

  /**
   * 根据用户名获取用户
   */
  async getByUsername(username) {
    try {
      const cacheKey = `${this.cacheKeyPrefix}username:${username}`;
      return getCache(cacheKey, async () => {
        const user = await db(this.tableName).where({ username }).first();
        return user || null;
      });
    } catch (error) {
      logger.error('根据用户名获取用户失败', { error: error.message, username });
      throw error;
    }
  }

  /**
   * 根据邮箱获取用户
   */
  async getByEmail(email) {
    try {
      const cacheKey = `${this.cacheKeyPrefix}email:${email}`;
      return getCache(cacheKey, async () => {
        const user = await db(this.tableName).where({ email }).first();
        return user || null;
      });
    } catch (error) {
      logger.error('根据邮箱获取用户失败', { error: error.message, email });
      throw error;
    }
  }

  /**
   * 根据手机号获取用户
   */
  async getByPhone(phone) {
    try {
      const cacheKey = `${this.cacheKeyPrefix}phone:${phone}`;
      return getCache(cacheKey, async () => {
        const user = await db(this.tableName).where({ phone }).first();
        return user || null;
      });
    } catch (error) {
      logger.error('根据手机号获取用户失败', { error: error.message, phone });
      throw error;
    }
  }

  /**
   * 创建用户
   */
  async create(userData) {
    try {
      const now = new Date();
      const user = {
        id: uuidv4(),
        ...userData,
        created_at: now,
        updated_at: now
      };

      await db(this.tableName).insert(user);
      
      // 清除缓存
      await this.clearUserCache(user);
      
      return user;
    } catch (error) {
      logger.error('创建用户失败', { error: error.message, userData });
      throw error;
    }
  }

  /**
   * 更新用户
   */
  async update(id, userData) {
    try {
      const now = new Date();
      const updateData = {
        ...userData,
        updated_at: now
      };

      await db(this.tableName).where({ id }).update(updateData);
      
      // 清除缓存
      await this.clearUserCache({ id, ...userData });
      
      return this.getById(id);
    } catch (error) {
      logger.error('更新用户失败', { error: error.message, id, userData });
      throw error;
    }
  }

  /**
   * 删除用户
   */
  async delete(id) {
    try {
      const user = await this.getById(id);
      
      if (!user) {
        return false;
      }
      
      await db(this.tableName).where({ id }).del();
      
      // 清除缓存
      await this.clearUserCache(user);
      
      return true;
    } catch (error) {
      logger.error('删除用户失败', { error: error.message, id });
      throw error;
    }
  }

  /**
   * 获取分页用户列表
   */
  async list(page = 1, pageSize = 10, filters = {}) {
    try {
      const offset = (page - 1) * pageSize;
      
      const query = db(this.tableName);
      
      // 应用过滤条件
      if (filters.role) {
        query.where('role', filters.role);
      }
      
      if (filters.status) {
        query.where('status', filters.status);
      }
      
      if (filters.verified !== undefined) {
        query.where('verified', filters.verified);
      }
      
      // 获取总数
      const countQuery = query.clone();
      const totalItems = await countQuery.count('id as count').first();
      
      // 获取分页数据
      const users = await query
        .select('*')
        .orderBy('created_at', 'desc')
        .offset(offset)
        .limit(pageSize);
      
      return {
        data: users,
        pagination: {
          page,
          pageSize,
          totalItems: totalItems ? totalItems.count : 0,
          totalPages: Math.ceil((totalItems ? totalItems.count : 0) / pageSize)
        }
      };
    } catch (error) {
      logger.error('获取用户列表失败', { error: error.message, page, pageSize, filters });
      throw error;
    }
  }

  /**
   * 更新用户最后登录时间
   */
  async updateLastLogin(id) {
    try {
      const now = new Date();
      await db(this.tableName).where({ id }).update({ last_login: now, updated_at: now });
      
      // 清除缓存
      await deleteCache(`${this.cacheKeyPrefix}id:${id}`);
      
      return true;
    } catch (error) {
      logger.error('更新用户最后登录时间失败', { error: error.message, id });
      throw error;
    }
  }

  /**
   * 验证用户
   */
  async verifyUser(id) {
    try {
      await db(this.tableName).where({ id }).update({
        verified: true,
        status: 'active',
        verification_token: null,
        verification_expires: null,
        updated_at: new Date()
      });
      
      // 清除缓存
      await deleteCache(`${this.cacheKeyPrefix}id:${id}`);
      
      return this.getById(id);
    } catch (error) {
      logger.error('验证用户失败', { error: error.message, id });
      throw error;
    }
  }

  /**
   * 更新验证令牌
   */
  async updateVerificationToken(id, token, expiresIn = 86400) {
    try {
      const now = new Date();
      const expiresAt = new Date(now.getTime() + expiresIn * 1000);
      
      await db(this.tableName).where({ id }).update({
        verification_token: token,
        verification_expires: expiresAt,
        updated_at: now
      });
      
      // 清除缓存
      await deleteCache(`${this.cacheKeyPrefix}id:${id}`);
      
      return true;
    } catch (error) {
      logger.error('更新验证令牌失败', { error: error.message, id });
      throw error;
    }
  }

  /**
   * 检查验证令牌
   */
  async checkVerificationToken(token) {
    try {
      const now = new Date();
      
      const user = await db(this.tableName)
        .where({ verification_token: token })
        .where('verification_expires', '>', now)
        .first();
      
      return user || null;
    } catch (error) {
      logger.error('检查验证令牌失败', { error: error.message, token });
      throw error;
    }
  }

  /**
   * 清除用户相关缓存
   */
  async clearUserCache(user) {
    try {
      if (!user) return;
      
      const cacheKeys = [];
      
      if (user.id) {
        cacheKeys.push(`${this.cacheKeyPrefix}id:${user.id}`);
      }
      
      if (user.username) {
        cacheKeys.push(`${this.cacheKeyPrefix}username:${user.username}`);
      }
      
      if (user.email) {
        cacheKeys.push(`${this.cacheKeyPrefix}email:${user.email}`);
      }
      
      if (user.phone) {
        cacheKeys.push(`${this.cacheKeyPrefix}phone:${user.phone}`);
      }
      
      if (cacheKeys.length > 0) {
        await Promise.all(cacheKeys.map(key => deleteCache(key)));
      }
      
      // 清除包含用户ID的所有缓存
      if (user.id) {
        await deleteCacheByPattern(`*:${user.id}:*`);
      }
      
      return true;
    } catch (error) {
      logger.error('清除用户缓存失败', { error: error.message, user });
      return false;
    }
  }
}

module.exports = new UserRepository(); 

File: Developer/suoke_life/services/auth-service/internal/repository/user_repository.go
package repository

import (
	"context"
	"database/sql"
	"errors"
	"fmt"
	"time"

	"github.com/jmoiron/sqlx"
	"github.com/suoke-life/auth-service/internal/models"
	"github.com/suoke-life/shared/pkg/logger"
)

// UserRepository 用户存储库接口
type UserRepository interface {
	Create(ctx context.Context, user *models.User) error
	GetByID(ctx context.Context, id string) (*models.User, error)
	GetByEmail(ctx context.Context, email string) (*models.User, error)
	GetByUsername(ctx context.Context, username string) (*models.User, error)
	Update(ctx context.Context, user *models.User) error
	UpdateLastLogin(ctx context.Context, userID string) error
	Delete(ctx context.Context, userID string) error
	List(ctx context.Context, offset, limit int) ([]*models.User, error)
	Count(ctx context.Context) (int, error)
}

// SQLUserRepository SQL用户存储库实现
type SQLUserRepository struct {
	db     *sqlx.DB
	logger logger.Logger
}

// 确保SQLUserRepository实现了UserRepository接口
var _ UserRepository = (*SQLUserRepository)(nil)

// NewSQLUserRepository 创建新的SQL用户存储库
func NewSQLUserRepository(db *sqlx.DB, log logger.Logger) *SQLUserRepository {
	return &SQLUserRepository{
		db:     db,
		logger: log.With("repository", "user"),
	}
}

// Create 创建新用户
func (r *SQLUserRepository) Create(ctx context.Context, user *models.User) error {
	query := `
		INSERT INTO users (
			id, username, email, password_hash, role, active, 
			email_verified, created_at, updated_at, last_login
		) VALUES (
			:id, :username, :email, :password_hash, :role, :active, 
			:email_verified, :created_at, :updated_at, :last_login
		)
	`

	user.CreatedAt = time.Now()
	user.UpdatedAt = user.CreatedAt

	_, err := r.db.NamedExecContext(ctx, query, user)
	if err != nil {
		r.logger.Error("创建用户失败", "error", err, "user", user.ID)
		return fmt.Errorf("创建用户失败: %w", err)
	}

	return nil
}

// GetByID 通过ID获取用户
func (r *SQLUserRepository) GetByID(ctx context.Context, id string) (*models.User, error) {
	query := `
		SELECT * FROM users WHERE id = $1
	`

	var user models.User
	err := r.db.GetContext(ctx, &user, query, id)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return nil, fmt.Errorf("未找到用户: %s", id)
		}
		r.logger.Error("获取用户失败", "error", err, "id", id)
		return nil, fmt.Errorf("获取用户失败: %w", err)
	}

	return &user, nil
}

// GetByEmail 通过邮箱获取用户
func (r *SQLUserRepository) GetByEmail(ctx context.Context, email string) (*models.User, error) {
	query := `
		SELECT * FROM users WHERE email = $1
	`

	var user models.User
	err := r.db.GetContext(ctx, &user, query, email)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return nil, fmt.Errorf("未找到用户: %s", email)
		}
		r.logger.Error("获取用户失败", "error", err, "email", email)
		return nil, fmt.Errorf("获取用户失败: %w", err)
	}

	return &user, nil
}

// GetByUsername 通过用户名获取用户
func (r *SQLUserRepository) GetByUsername(ctx context.Context, username string) (*models.User, error) {
	query := `
		SELECT * FROM users WHERE username = $1
	`

	var user models.User
	err := r.db.GetContext(ctx, &user, query, username)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return nil, fmt.Errorf("未找到用户: %s", username)
		}
		r.logger.Error("获取用户失败", "error", err, "username", username)
		return nil, fmt.Errorf("获取用户失败: %w", err)
	}

	return &user, nil
}

// Update 更新用户信息
func (r *SQLUserRepository) Update(ctx context.Context, user *models.User) error {
	query := `
		UPDATE users
		SET username = :username,
			email = :email,
			password_hash = :password_hash,
			role = :role,
			active = :active,
			email_verified = :email_verified,
			updated_at = :updated_at
		WHERE id = :id
	`

	user.UpdatedAt = time.Now()

	_, err := r.db.NamedExecContext(ctx, query, user)
	if err != nil {
		r.logger.Error("更新用户失败", "error", err, "user", user.ID)
		return fmt.Errorf("更新用户失败: %w", err)
	}

	return nil
}

// UpdateLastLogin 更新用户最后登录时间
func (r *SQLUserRepository) UpdateLastLogin(ctx context.Context, userID string) error {
	query := `
		UPDATE users
		SET last_login = $1
		WHERE id = $2
	`

	now := time.Now()
	_, err := r.db.ExecContext(ctx, query, now, userID)
	if err != nil {
		r.logger.Error("更新用户最后登录时间失败", "error", err, "user", userID)
		return fmt.Errorf("更新用户最后登录时间失败: %w", err)
	}

	return nil
}

// Delete 删除用户
func (r *SQLUserRepository) Delete(ctx context.Context, userID string) error {
	query := `
		DELETE FROM users WHERE id = $1
	`

	_, err := r.db.ExecContext(ctx, query, userID)
	if err != nil {
		r.logger.Error("删除用户失败", "error", err, "user", userID)
		return fmt.Errorf("删除用户失败: %w", err)
	}

	return nil
}

// List 列出用户
func (r *SQLUserRepository) List(ctx context.Context, offset, limit int) ([]*models.User, error) {
	query := `
		SELECT * FROM users
		ORDER BY created_at DESC
		LIMIT $1 OFFSET $2
	`

	users := []*models.User{}
	err := r.db.SelectContext(ctx, &users, query, limit, offset)
	if err != nil {
		r.logger.Error("列出用户失败", "error", err)
		return nil, fmt.Errorf("列出用户失败: %w", err)
	}

	return users, nil
}

// Count 获取用户总数
func (r *SQLUserRepository) Count(ctx context.Context) (int, error) {
	query := `
		SELECT COUNT(*) FROM users
	`

	var count int
	err := r.db.GetContext(ctx, &count, query)
	if err != nil {
		r.logger.Error("获取用户总数失败", "error", err)
		return 0, fmt.Errorf("获取用户总数失败: %w", err)
	}

	return count, nil
} 